### PR TITLE
Table: Indexes & Joins

### PR DESCRIPTION
**Pull request by kustosz**
### Pull Request Description
Introduces indexes and simple joins on the table structure.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: table/src/main/java/org/enso/table/data/table/Table.java [hunk 1]
```diff
@@ -99,13 +115,136 @@ private Table replaceColumn(int ix, Column newCol) {
...
+        countMask[i] = matches[i].size();
+      }
+      outSize += countMask[i];
+    }
+    int[] orderMask = new int[outSize];
+    int orderMaskPosition = 0;
+    for (int i = 0; i < s; i++) {
+      if (matches[i] == null) {
+        if (!dropUnmatched) {
+          orderMask[orderMaskPosition++] = -1;
```
Comment (radeusgd) [inline table/src/main/java/org/enso/table/data/table/Table.java]:
```suggestion
          orderMask[orderMaskPosition++] = Index.NOT_FOUND;
```

**Review by radeusgd — APPROVED** [review]
Looks good to me.

Some minor suggestions inline.

One important thing I would like to clarify is handling of these ANSI terminal codes as this may be potentially problematic on Windows, was this tested? It will also cause trouble if piping such output to file (although of course this output format is not meant for files, I bet someone will try saving it to a file anyway).

File: distribution/std-lib/Table/src/Data/Column.enso [hunk 2]
```diff
@@ -13,13 +13,16 @@ type Column
     display : Integer -> Text
     display show_rows=10 =
         java_col = this.java_column
+        index = this.java_column.getIndex []
```
[the remark this answers is not in the document]
Comment (kustosz) [reply distribution/std-lib/Table/src/Data/Column.enso]: I don't understand the question, can you rephrase?

[the remark this answers is not in the document]
Comment (radeusgd) [reply distribution/std-lib/Table/src/Data/Column.enso]: Sorry, my sentence generation engine must have crahsed.

What I meant is, why don't you use `java_col` in this line like in all the other lines but instead refer to the full `this.java_column` which is equivalent to `java_col` here. Just asking because it was inconsistent.

File: table/src/main/java/org/enso/table/data/table/Table.java [hunk 3]
```diff
@@ -99,13 +115,136 @@ private Table replaceColumn(int ix, Column newCol) {
...
+   * Selects a subset of columns of this table, by names.
+   *
+   * @param colNames the column names to select
+   * @return a table containing only selected columns
+   */
+  public Table selectColumns(List<String> colNames) {
+    Column[] newCols =
+        colNames.stream()
+            .map(this::getColumnByName)
+            .filter(Objects::nonNull)
```
[the remark this answers is not in the document]
Comment (kustosz) [reply table/src/main/java/org/enso/table/data/table/Table.java:168]: Nah, I'd rather be permissive here
